Plugins/Data/NWScript/inc_array.nss

```diff
@@ -0,0 +1,485 @@
+        }
+    }
+    return FALSE;
+}
+
+int Array_Contains_Flt(string tag, float element, object obj=OBJECT_INVALID)
+{
+    return Array_Contains_Str(tag, FloatToString(element), obj);
+}
+
+int Array_Contains_Int(string tag, int element, object obj=OBJECT_INVALID)
+{
+    return Array_Contains_Str(tag, IntToString(element), obj);
+}
+
+int Array_Contains_Obj(string tag, object element, object obj=OBJECT_INVALID)
+{
+    return Array_Contains_Str(tag, ObjectToString(element), obj);
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+void Array_Copy(string tag, string otherTag, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    CreateArrayTable(otherTag, obj);
+    ExecuteStatement("insert into "+GetTableName(otherTag, obj)+" select * from "+GetTableName(tag, obj), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+void Array_Erase(string tag, int index, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    ExecuteStatement("DELETE FROM "+GetTableName(tag, obj)+" WHERE ind = " + IntToString(index), obj);
+    ExecuteStatement("UPDATE "+GetTableName(tag, obj)+" SET ind = ind - 1 WHERE ind > " + IntToString(index), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// return the index in the array containing "element"
+// if not found, return INVALID_INDEX
+int Array_Find_Str(string tag, string element, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    int tmp = -1;
+    string stmt = "SELECT MIN(ind) FROM "+GetTableName(tag, obj)+" WHERE value = @element";
+    sqlquery sqlQuery = SqlPrepareQueryObject(GetModule(), stmt);
+    SqlBindString(sqlQuery, "@element", element);
+    if ( SqlStep(sqlQuery) ) {
+        tmp = SqlGetInt(sqlQuery, 0);
+        WriteTimestampedLogEntry("Returned value = " + IntToString(tmp));
+        return tmp;
+    }
+    return INVALID_INDEX;
+}
+
+int Array_Find_Flt(string tag, float element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, FloatToString(element), obj);
+}
+
+int Array_Find_Int(string tag, int element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, IntToString(element), obj);
+}
+
+int Array_Find_Obj(string tag, object element, object obj=OBJECT_INVALID)
+{
+    return Array_Find_Str(tag, ObjectToString(element), obj);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Insert a new element into position 'index'.  If index is beyond the number of rows in the array,
+// this will quietly fail.  This could be changed if you wanted to support sparse
+// arrays.
+void Array_Insert_Str(string tag, int index, string element, object obj=OBJECT_INVALID)
+{
+    int rows = GetRowCount(tag, obj);
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    if (index < rows) {
```

Comment:
Need also to check for `index >= 0`; Also `index==rows` is valid if you want to insert at the end.